src/gc.c

```diff
@@ -249,6 +249,8 @@ static void gc_add_finalizer_(jl_ptls_t ptls, void *v, void *f)
     size_t oldlen = jl_atomic_load_acquire(&a->len);
     if (__unlikely(oldlen + 2 > a->max)) {
         JL_LOCK_NOGC(&finalizers_lock);
+        // `a->len` might have been modified.
```

Comment:
It doesn't hurt to call `arraylist_grow` when the length is shrunk by another thread already (the grow will just not allocate any memory).